File: day5/p1.cpp
/*
- Parse the rules into a dictionary.
- Validate updates sequentially.
- Calculate the middle page for each valid update.
- Compute the sum.
*/

#include <iostream>
#include <fstream>
#include <sstream>
#include <vector>
#include <unordered_map>
#include <set>
#include <algorithm>

using namespace std;

// Function to parse input from file
void parseInput(const string& filename, unordered_map<int, set<int>>& rules, vector<vector<int>>& updates) {
    ifstream file(filename);
    if (!file) {
        cerr << "Error opening file!" << endl;
        return;
    }

    string line;
    bool parsingRules = true;

    while (getline(file, line)) {
        if (line.empty()) {
            parsingRules = false; // Switch to parsing updates
            continue;
        }

        if (parsingRules) {
            // Parse rules (e.g., "X|Y")
            stringstream ss(line);
            int x, y;
            char delim;
            ss >> x >> delim >> y;
            rules[x].insert(y);
        } else {
            // Parse updates (e.g., "75,47,61,53,29")
            stringstream ss(line);
            vector<int> update;
            int page;
            char delim;

            while (ss >> page) {
                update.push_back(page);
                ss >> delim; // Skip the comma
            }
            updates.push_back(update);
        }
    }
}

// Function to validate an update against rules
bool isValidUpdate(const vector<int>& update, const unordered_map<int, set<int>>& rules) {
    unordered_map<int, int> positions;

    // Map each page to its position in the update
    for (int i = 0; i < update.size(); ++i) {
        positions[update[i]] = i;
    }

    // Validate rules
    for (const auto& rule : rules) {
        int x = rule.first;
        for (int y : rule.second) {
            // If both x and y are in the update, check their order
            if (positions.count(x) && positions.count(y)) {
                if (positions[x] > positions[y]) {
                    return false;
                }
            }
        }
    }
    return true;
}

// Function to find the middle page of an update
int findMiddlePage(const vector<int>& update) {
    int n = update.size();
    return update[n / 2];
}

int main() {
    string filename = "input_puzzle.txt";

    unordered_map<int, set<int>> rules; // Stores rules as a map of X -> {Y}
    vector<vector<int>> updates;       // Stores updates as a vector of vectors

    // Parse the input file
    parseInput(filename, rules, updates);

    // Process each update
    int sumMiddlePages = 0;
    for (const auto& update : updates) {
        if (isValidUpdate(update, rules)) {
            sumMiddlePages += findMiddlePage(update);
        }
    }

    // Output the result
    cout << "Sum of middle pages: " << sumMiddlePages << endl;

    return 0;
}

File: day_5/p1.js
const fs = require("fs");

// Function to parse input from file
function parseInput(filename) {
    const rules = new Map();
    const updates = [];

    try {
        const data = fs.readFileSync(filename, "utf8");
        const lines = data.split("\n");

        let parsingRules = true;

        lines.forEach(line => {
            line = line.trim();
            if (!line) {
                parsingRules = false; // Switch to parsing updates
                return;
            }

            if (parsingRules) {
                // Parse rules (e.g., "X|Y")
                const [x, y] = line.split("|").map(Number);
                if (!rules.has(x)) {
                    rules.set(x, new Set());
                }
                rules.get(x).add(y);
            } else {
                // Parse updates (e.g., "75,47,61,53,29")
                const update = line.split(",").map(Number);
                updates.push(update);
            }
        });
    } catch (err) {
        console.error("Error reading file:", err.message);
    }

    return { rules, updates };
}

// Function to validate an update against rules
function isValidUpdate(update, rules) {
    const positions = new Map();

    // Map each page to its position in the update
    update.forEach((page, index) => {
        positions.set(page, index);
    });

    // Validate rules
    for (const [x, ys] of rules.entries()) {
        for (const y of ys) {
            // If both x and y are in the update, check their order
            if (positions.has(x) && positions.has(y)) {
                if (positions.get(x) > positions.get(y)) {
                    return false;
                }
            }
        }
    }
    return true;
}

// Function to find the middle page of an update
function findMiddlePage(update) {
    const n = update.length;
    return update[Math.floor(n / 2)];
}

// Main function
function main() {
    const filename = "input_puzzle.txt";

    // Parse the input file
    const { rules, updates } = parseInput(filename);

    // Process each update
    let sumMiddlePages = 0;
    updates.forEach(update => {
        if (isValidUpdate(update, rules)) {
            sumMiddlePages += findMiddlePage(update);
        }
    });

    // Output the result
    console.log("Sum of middle pages:", sumMiddlePages);
}

// Execute the main function
main();
